pinot-broker/src/test/java/com/linkedin/pinot/broker/routing/builder/KafkaHighLevelConsumerRoutingTableBuilderTest.java

```diff
@@ -0,0 +1,114 @@
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.InstanceConfig;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+
+public class KafkaHighLevelConsumerRoutingTableBuilderTest {
+
+  public static final String ALL_PARTITIONS = "ALL";
+
+  @Test
+  public void testHlcRoutingTableBuilder() {
+    final int ITERATIONS = 50;
+    final int MAX_INSTANCES_PER_GROUP = 5;
+    final int MAX_NUM_GROUPS = 5;
+    final int MAX_NUM_SEGMENTS = 100;
+
+    Random random = new Random();
+
+    KafkaHighLevelConsumerBasedRoutingTableBuilder routingTableBuilder =
+        new KafkaHighLevelConsumerBasedRoutingTableBuilder();
+    routingTableBuilder.init(new BaseConfiguration(), new TableConfig(), null, null);
+
+    String tableName = "table_REALTIME";
+    String groupPrefix = "table_REALTIME_" + System.currentTimeMillis();
+
+    // Generate instances
+    for (int i = 0; i < ITERATIONS; i++) {
+      int numInstancesPerGroup = random.nextInt(MAX_INSTANCES_PER_GROUP) + 1;
+      int numGroups = random.nextInt(MAX_NUM_GROUPS) + 1;
+      int numSegments = random.nextInt(MAX_NUM_SEGMENTS) + 5;
+      int instanceCount = numInstancesPerGroup * numGroups;
+
+      // Generate instance names
+      String[] instanceNames = new String[instanceCount];
+      for (int serverInstanceId = 0; serverInstanceId < instanceCount; serverInstanceId++) {
+        instanceNames[serverInstanceId] = "Server_localhost_" + serverInstanceId;
+      }
+
+      // Generate instance configurations
+      List<InstanceConfig> instanceConfigs = new ArrayList<>();
+      for (String instanceName : instanceNames) {
+        InstanceConfig instanceConfig = new InstanceConfig(instanceName);
+        instanceConfigs.add(instanceConfig);
+        instanceConfig.getRecord().setSimpleField(CommonConstants.Helix.IS_SHUTDOWN_IN_PROGRESS, "false");
+      }
+
+      // Generate fake external view
+      ExternalView externalView = new ExternalView(tableName);
+      Map<String, Set<String>> groupIdToSegmentsMapping = new HashMap<>();
+      for (int segmentNum = 0; segmentNum < numSegments; segmentNum++) {
+        int groupId = segmentNum % numGroups;
+        int groupOffset = segmentNum % numInstancesPerGroup;
+        String instanceName = instanceNames[groupId * numInstancesPerGroup + groupOffset];
+        String groupName = groupPrefix + "_" + groupId;
+        String segmentName = new HLCSegmentName(groupName, ALL_PARTITIONS, String.valueOf(segmentNum)).getSegmentName();
+        externalView.setState(segmentName, instanceName, "ONLINE");
+
+        Set<String> segmentsForGroupId = groupIdToSegmentsMapping.computeIfAbsent(groupName, key -> new HashSet<>());
+        segmentsForGroupId.add(segmentName);
+      }
+
+      // Compute routing table
+      routingTableBuilder.computeRoutingTableFromExternalView(tableName, externalView, instanceConfigs);
+
+      // Check if the routing table result is correct
+      for (int run = 0; run < 30; run++) {
```

Comment:
You can also put `30` into a constant.

Reply:
changed to `NUM_GROUPS (number of routing tables) * 10`